google/cloud/ndb/_cache.py

```diff
@@ -304,7 +304,33 @@ class _GlobalCacheSetBatch(_GlobalCacheBatch):
     def __init__(self, options):
         self.expires = options.get("expires")
         self.todo = {}
-        self.futures = []
+        self.futures = {}
+
+    def done_callback(self, cache_call):
+        """Process results of call to global cache.
+        If there is an exception for the cache call, distribute that to waiting
+        futures, otherwise examine the result of the cache call. If the result is
+        :data:`None`, simply set the result to :data:`None` for all waiting futures.
+        Otherwise, if the result is a `dict`, use that to propagate results for
+        individual keys to waiting figures.
+        """
+        exception = cache_call.exception()
+        if exception:
+            for future in self.futures.values():
+                future.set_exception(exception)
+            return
+
+        result = cache_call.result()
+        if result:
+            for key, future in self.futures.items():
+                key_result = result.get(key, None)
```

Comment:
Should the default value be None or False in the case of `set_if_not_exists`?

Reply:
I don't think so. In the case of `set_if_not_exists`, there will be a result for the key, so the default won't be used.

Reply:
To me it's more of an api contractual thing.  Both setnx and memcache.add return true if it was set and false if it was not.  So what does None mean?

Reply:
This code is also used by `set` not just `set_if_not_exists`. In the case of `set_if_not_exists`, it will always follow the path that eventually returns a boolean value. `None` won't ever be returned for `set_if_not_exists`.